File: pkg/gotth/live/devreload.go
package live

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"

	"github.com/a-h/templ"
)

// clientDevReloadFile is the dev-reload client, served and pointed at only
// when Config.Dev is set. It is a THIRD artifact for the reason the inspector
// is a second one: FR-57 is a development-time convenience and NFR-2's
// 12,288-byte ceiling is a production budget, so not one byte of this may be
// spent inside the shipped runtime. client/SIZE.md §2.2 carries the
// measurement and tools/minify holds it to its own ceiling.
const clientDevReloadFile = "gotth-live-dev-reload.min.js"

// devBuildRoute is the path suffix the dev-reload client polls for the
// identity of the running build.
//
// It carries no file extension on purpose. Both JavaScript artifacts are
// routed by ".min.js" suffix, and a fourth name in that family would be a
// fourth chance for the suffix tests to overlap; this one cannot collide with
// any of them however they are renamed.
const devBuildRoute = "gotth-live-dev-build"

// maxDevBuildID bounds Config.DevBuildID.
//
// The value is written into an HTML attribute and returned as an HTTP body, so
// the bound is not a formality: client/dev-reload.js refuses a body longer
// than this rather than treating it as a build identity, because a long body
// on this route is far more likely to be a proxy's error page than a build.
// The two numbers are the same number, and this is the one a developer can
// see fail at startup.
const maxDevBuildID = 128

// clientDevReload is the minified dev-reload client, embedded by exact
// filename for the same reason the other two are.
//
// Embedding is unconditional and SERVING is not, exactly as for the inspector
// — see clientInspector, which makes the whole argument. A production binary
// carries these bytes and hands them to nobody.
//
//go:embed clientjs/gotth-live-dev-reload.min.js
var clientDevReload []byte

// clientDevReloadETag is computed once, at init, like the others'.
var clientDevReloadETag = etagOf(clientDevReload)

// DevReloadScript renders the script tag for the dev-reload client, for an
// application whose handler is mounted at mountPath.
//
// It is the browser half of FR-57: after a Go or templ change is rebuilt and
// the process restarts, the page in front of the developer reloads itself.
// docs/guide/dev-reload.md is the user-facing page.
//
// # What it does, and what the runtime already does
//
// A Go change and a templ change are one event — templ generates Go — so both
// mean a rebuild and a restart. The socket drops either way, and the client
// runtime's reconnect-and-resync brings the live regions back on its own; that
// needs nothing from this tag, and it is why restarting the SAME binary does
// not reload anything here.
//
// What a resync cannot repaint is everything outside a live fragment: the page
// shell, the head, a fragment whose markup changed while its state did not and
// which therefore produced no patch. After a rebuild that markup came from a
// process that no longer exists. This tag polls the build identity below and
// reloads the document when it changes, which is the only thing that fixes it.
//
// # What it does NOT preserve
//
// Server-held session state does not survive the process that held it. The
// reconnect mounts a NEW session against the new build, so Config.Init runs
// again and the session's state is whatever Init produces. That is not
// something this tag could preserve and it does not pretend to: "without
// losing the session where state permits" means the browser re-establishes
// itself with no manual refresh, not that a restarted process remembers.
// Application state kept outside the session — the counter example's store is
// the one to look at — survives exactly as far as its own lifetime allows,
// which for an in-process store is not at all.
//
// # It renders nothing unless Config.Dev is set
//
// With Dev false — the zero value, and what production must run — this writes
// zero bytes and returns nil, the route serving its JavaScript answers 404,
// and so does the build-identity route. Three gates on one switch, all three
// tested in both positions. A production page therefore names no dev asset,
// exposes no build identity, and makes no polling request.
//
// # Order does not matter here
//
// Unlike (*App[S, I]).InspectorScript, this tag may go anywhere: it wraps
// nothing, reads nothing the runtime owns, and talks only to its own route
// over HTTP. Putting all three tags together, inspector first and this one
// last, is the arrangement the guide shows, and only the inspector's position
// in it is load-bearing.
//
// mountPath is validated exactly as Script validates it, by the same function,
// and for the same reason: the prefix as the browser sees it is knowledge only
// the caller has.
func (a *App[S, I]) DevReloadScript(mountPath string) templ.Component {
	return templ.ComponentFunc(func(_ context.Context, w io.Writer) error {
		// Validated even when nothing is written, so a bad mount path is an
		// error in dev and in production alike — the argument InspectorScript
		// makes at length, and the same function makes the decision.
		mount, err := normalizeMount(mountPath)
		if err != nil {
			return err
		}
		if !a.cfg.Dev {
			return nil
		}
		src := mount + "/" + clientDevReloadFile
		if mount == "/" {
			src = mount + clientDevReloadFile
		}
		// The build identity is written into the tag rather than fetched by
		// the client at boot, and that is the correctness of the feature
		// rather than a saved round trip: this is the identity of the build
		// that rendered THIS document. A client that adopted its first fetched
		// value as the baseline would accept a rebuild that landed while the
		// page was loading as "what this page is showing", and never reload.
		_, err = io.WriteString(w,
			`<script src="`+templ.EscapeString(src)+`" `+
				attrDevURL+`="`+templ.EscapeString(mount)+`" `+
				attrDevBuild+`="`+templ.EscapeString(a.devBuildID())+`" defer></script>`)
		return err
	})
}

// devBuildID is the identity of the running build.
//
// Config.DevBuildID wins when it is set. Otherwise it is derived, once per
// process, from the bytes of the running executable — see devBuildID's
// package-level helper below for what that buys and what it costs.
func (a *App[S, I]) devBuildID() string {
	if a.cfg.DevBuildID != "" {
		return a.cfg.DevBuildID
	}
	a.buildOnce.Do(func() { a.buildID = executableBuildID() })
	return a.buildID
}

// executableBuildID hashes the running executable.
//
// # Why the executable, and not a per-process random value
//
// Both change across a rebuild, which is all FR-57 strictly needs. The hash
// additionally does NOT change across a restart that rebuilt nothing, and that
// difference is the whole of "without losing the session where state permits":
// a crash-and-restart, a `docker compose restart`, or a rebuild of source that
// did not actually change all leave the identity alone, so the page does not
// reload and the client runtime's own reconnect restores it in place. Go
// builds are reproducible, so "did not actually change" is a property the
// toolchain gives for free here.
//
// It costs one streaming SHA-256 of the binary, taken lazily on the first poll
// rather than at startup, once per process, in dev only. A 40 MB development
// binary is a few tens of milliseconds on the first request and nothing
// afterwards.
//
// # The fallback, and why it is visibly named
//
// If the executable cannot be found or read — a binary deleted out from under
// a running process, a platform where os.Executable is unsupported — the
// identity becomes a random per-process value prefixed "process-". That is the
// safe direction: every restart then looks like a new build and reloads the
// page, which is a page that reloads more often than it must rather than one
// that shows stale markup. The prefix is there so a developer reading
// docs/guide/dev-reload.md can tell which of the two they are getting.
func executableBuildID() string {
	if path, err := os.Executable(); err == nil {
		if f, err := os.Open(path); err == nil {
			defer f.Close()
			sum := sha256.New()
			if _, err := io.Copy(sum, f); err == nil {
				return hex.EncodeToString(sum.Sum(nil)[:12])
			}
		}
	}
	var b [12]byte
	if _, err := rand.Read(b[:]); err == nil {
		return "process-" + hex.EncodeToString(b[:])
	}
	// rand.Read does not return an error on any supported platform; the arm
	// exists so that this function has no path that returns "".  An empty
	// identity would be rendered into the tag as an empty attribute, which
	// client/dev-reload.js reads as "no baseline" and quietly stops watching.
	return "process-unavailable"
}

// serveClientDevReload serves the embedded dev-reload client.
//
// Reached only from the Dev arm of routes(), and served with the same
// immutable caching as the other two artifacts: it changes only when the
// binary does, and a developer reloading fifty times should not fetch it
// fifty times.
func serveClientDevReload(w http.ResponseWriter, r *http.Request) {
	serveAsset(w, r, clientDevReload, clientDevReloadETag)
}

// serveDevBuild answers with the identity of the running build.
//
// It is the one route in this library that must never be cached, and it is
// deliberately not served through serveAsset: that function's whole job is a
// strong ETag and a year of immutability, which is the exact opposite of what
// this answer needs. A conditional request is not offered either — the answer
// is 12 bytes, and a 304 that a proxy decided to satisfy from a store is a
// page that never reloads.
func (a *App[S, I]) serveDevBuild(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet && r.Method != http.MethodHead {
		w.Header().Set("Allow", "GET, HEAD")
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	id := a.devBuildID()
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Header().Set("Content-Length", strconv.Itoa(len(id)))
	if r.Method == http.MethodHead {
		return
	}
	_, _ = io.WriteString(w, id)
}

// validateDevBuildID checks a Config.DevBuildID a caller set.
//
// It is checked whatever Dev is set to, for the reason InspectorScript
// validates its mount path in both modes: a field that is only checked in one
// of them is a field that starts failing on the deploy that flips the switch.
func validateDevBuildID(id string) error {
	if id == "" {
		return nil
	}
	if len(id) > maxDevBuildID {
		return &ConfigError{
			Field: "DevBuildID",
			Detail: "the value is " + strconv.Itoa(len(id)) + " bytes and the limit is " +
				strconv.Itoa(maxDevBuildID) + ": it is rendered into a script tag and returned as the " +
				"whole body of the dev-reload poll, and the client refuses a longer answer rather than " +
				"treating it as a build identity — a short opaque token such as a commit hash is what fits",
		}
	}
	for i := 0; i < len(id); i++ {
		if b := id[i]; b < 0x20 || b == 0x7f {
			return &ConfigError{
				Field: "DevBuildID",
				Detail: "the value contains the control byte " + strconv.QuoteRune(rune(b)) +
					": it is returned as a single-line HTTP body and compared verbatim in the browser, " +
					"so a newline or a tab in it makes the comparison depend on how a proxy rewrote it",
			}
		}
	}
	if strings.TrimSpace(id) != id {
		return &ConfigError{
			Field: "DevBuildID",
			Detail: "the value has leading or trailing whitespace: the client trims the body it " +
				"receives before comparing, so this identity can never equal itself — trim it here",
		}
	}
	return nil
}

File: pkg/gotth/client/dev-reload.js
// gotth-live dev reload — the browser half of FR-57.
//
// This file is NOT part of the shipped runtime. It is built to its own
// artifact, live/clientjs/gotth-live-dev-reload.min.js, served only when
// live.Config.Dev is true, and referenced only by the tag
// (*live.App).DevReloadScript renders — which renders nothing at all when Dev
// is false. It is the same two-gates-on-one-switch arrangement the inspector
// uses, for the same reason, and it costs the NFR-2 bundle exactly zero bytes:
// nothing in client/runtime.js knows this file exists, and nothing here
// imports it.
//
// ---------------------------------------------------------------------------
// What problem this solves, and what the runtime already solves
// ---------------------------------------------------------------------------
//
// A Go change and a templ change are the same event: templ generates Go, so
// both mean the process is rebuilt and restarted. The socket drops, and the
// client runtime's own reconnect-and-resync state machine brings the page back
// by itself — that part needs nothing from this file, and it is why a restart
// of the SAME binary is not a reload here.
//
// What resync cannot do is repaint anything outside a live fragment. The page
// shell, the <head>, the classes on the <body>, a fragment whose markup
// changed while its state did not (Dirty says clean, so no patch is sent) —
// all of that was rendered once, by the process that is now gone. After a
// rebuild the browser is showing markup from a build that no longer exists,
// with a live connection into a build that renders it differently. Only a
// document reload fixes that, and FR-57 says the developer must not be the one
// to ask for it.
//
// So this file answers one question, once a second: is the process on the
// other end still the build that rendered this page?
//
// ---------------------------------------------------------------------------
// Build identity, and why it is on the script tag
// ---------------------------------------------------------------------------
//
// The server exposes a build identity at <mount>/gotth-live-dev-build, in dev
// only. By default it is a hash of the running executable, so it changes when
// and only when the code changes: an identical rebuild, or a plain restart of
// the same binary, keeps it — and this file then does nothing at all, leaving
// the runtime's reconnect to restore the session.
//
// The baseline comes from the SCRIPT TAG, not from a first fetch at boot. That
// is not a micro-optimisation, it is the correctness of the whole thing: the
// server renders the id of the build that produced THIS document into the tag,
// so a rebuild that lands between the page render and the first poll is still
// caught. Adopting the first fetched value as the baseline would silently
// accept the new build as "what this page is showing", which is exactly the
// case a developer hits when they save a file while the page is loading.
//
// ---------------------------------------------------------------------------
// What is exported, and why
// ---------------------------------------------------------------------------
//
// The decision — reload, or keep waiting, or carry on — is pure, and it is
// exported so client/test/dev-reload.test.mjs can drive it directly rather
// than through a browser. check() is exported for the same reason one layer
// out: the mapping from "what fetch did" to "what that means" is where a
// non-2xx answer, a network error and a body of the wrong shape have to be
// told apart, and every one of those is a real thing a half-restarted server
// does. Importing this module under node runs no boot code, the same way
// runtime.js and inspector.js guard theirs.

//#region model

// The poll cadence, in milliseconds.
//
// STEADY is the healthy case: the server answered and the build matched. One
// request a second per open dev tab is the cost of the feature, and it is why
// polling stops while the tab is hidden.
//
// REBUILDING is the cadence while the server is not answering, which is what a
// rebuild-and-restart looks like from here. It is four times a second so the
// reload lands as soon as the new process is listening rather than up to a
// second later.
//
// The two decays after that exist because "not answering" also describes a
// server the developer stopped an hour ago, and hammering a dead port at 4 Hz
// until the tab is closed is rude to nothing except the log, but it is still
// wrong. After ten seconds of silence the cadence drops to STEADY, and after
// about a minute and a half to ABANDONED.
var STEADY = 1000;
var REBUILDING = 250;
var ABANDONED = 5000;

// The miss counts at which the cadence decays. Named rather than inlined
// because client/test/dev-reload.test.mjs asserts the boundaries by number,
// and a table whose numbers are written twice is a table that disagrees.
var REBUILDING_UNTIL = 40; // 40 × 250 ms = 10 s
var STEADY_UNTIL = 100; // then 60 × 1000 ms = 1 min more

// A build identity is a short opaque token. The bound is here rather than only
// on the server because this file must decide what to do with a body it did
// not expect, and "reload the page" is the wrong answer to a proxy's HTML
// error document that happened to arrive with a 200.
var MAX_BUILD_BYTES = 128;

var cfg = {
  // build is the identity of the build that rendered this document, read off
  // the script tag. Null means the tag carried none, which is a server that
  // is not actually in dev mode; nothing is polled in that case.
  build: null,
  // endpoint is the absolute-path URL of the build-identity route.
  endpoint: null,
};

var misses = 0;
var reloading = false;

// configure sets the baseline and the endpoint. It is called once at boot from
// the script tag's attributes, and directly by the specs.
export function configure(options) {
  cfg.build = options && typeof options.build === "string" && options.build !== "" ? options.build : null;
  cfg.endpoint = (options && options.endpoint) || null;
  misses = 0;
  reloading = false;
}

// reset returns the module to its unconfigured state. Specs only.
export function reset() {
  configure(null);
}

// state is the model the specs read, and what the badge renders from.
export function state() {
  return { build: cfg.build, endpoint: cfg.endpoint, misses: misses, reloading: reloading };
}

// observe folds one poll result into the model and returns what to do about
// it. It is the whole decision, and it is pure.
//
// The three answers:
//
//	"linked"    the build on the other end is the one that rendered this page.
//	            Nothing to do. This is the answer on every poll of a healthy
//	            session, and also the answer while the process is restarting
//	            into the SAME binary — that case is the runtime's reconnect to
//	            handle, not this file's.
//	"waiting"   nobody answered, or what answered was not a build identity.
//	            Between a rebuild's `kill` and its `listen` this is the answer
//	            for as many polls as the build takes.
//	"reload"    a DIFFERENT build answered. The document in front of the
//	            developer was rendered by a process that no longer exists.
//
// Once "reload" has been returned it is returned forever: location.reload() is
// asynchronous, several polls can land before the document goes away, and
// reloading a page that is already reloading is how a dev tool turns a rebuild
// into a loop.
export function observe(result) {
  if (reloading) return "reload";
  if (!result || result.ok !== true || typeof result.build !== "string") {
    misses++;
    return "waiting";
  }
  var build = result.build;
  if (build === "" || build.length > MAX_BUILD_BYTES) {
    // A 200 carrying something that is not a build identity is not evidence
    // that the build changed. It is evidence that something else answered —
    // a proxy, a login page, a router that swallowed the path — and the safe
    // reading of "I do not know" is "do not reload".
    misses++;
    return "waiting";
  }
  if (cfg.build === null) {
    // No baseline was rendered into the tag, so there is nothing to compare
    // against and adopting this one would make the first rebuild invisible.
    // Recorded, and nothing else.
    cfg.build = build;
    misses = 0;
    return "linked";
  }
  if (build !== cfg.build) {
    reloading = true;
    return "reload";
  }
  misses = 0;
  return "linked";
}

// nextDelay is how long to wait before polling again, given how many
// consecutive polls have failed. See STEADY/REBUILDING/ABANDONED above.
export function nextDelay() {
  if (misses === 0) return STEADY;
  if (misses <= REBUILDING_UNTIL) return REBUILDING;
  if (misses <= STEADY_UNTIL) return STEADY;
  return ABANDONED;
}

// buildURL turns the mount the server rendered into the tag into the path this
// file polls.
//
// The root mount is the one case: "/" already ends in the separator, and
// joining it again would spell "//" — which a browser reads as the beginning
// of an authority, so the poll would go to a host called
// "gotth-live-dev-build". live.normalizeMount refuses that spelling on the
// server for the same reason, and this is the client's half of it.
export function buildURL(mount) {
  return (mount === "/" ? "/" : mount + "/") + "gotth-live-dev-build";
}

//#endregion

//#region poll

// check performs one poll and folds the answer in.
//
// Everything a half-restarted server can do arrives here as an exception or as
// a response, and the mapping between those and observe()'s vocabulary is the
// part worth specifying: a refused connection throws, a listening socket
// belonging to something else answers 404 or 502, and a server that is up
// answers 200 with a short token. Only the last of those is evidence about the
// build.
//
// The fetch is `cache: "no-store"`: the answer is by definition the one thing
// on the page that must never be served from a cache, and the route sets
// no-store as well. Both, because either alone has been enough to make this
// class of feature quietly stop working.
export async function check(fetchImpl) {
  if (!cfg.endpoint) return observe(null);
  try {
    var response = await fetchImpl(cfg.endpoint, { cache: "no-store", credentials: "same-origin" });
    if (!response || !response.ok) return observe(null);
    var body = await response.text();
    return observe({ ok: true, build: body.trim() });
  } catch (err) {
    return observe(null);
  }
}

//#endregion

//#region badge

// The badge is the one piece of UI here: a small fixed marker that appears
// while the server is not answering and goes away when it comes back. It
// exists because the difference between "my rebuild is taking eight seconds"
// and "my app crashed and nothing is coming" is otherwise invisible — the page
// just sits there looking fine, which is the failure mode this project cares
// about most.
//
// It is built the way the inspector's panel is built, and for the same reason:
// CSSOM property writes and a constructed stylesheet, never a <style> element
// and never a style attribute, so it opens under a strict CSP with no
// 'unsafe-inline' and no nonce to offer. Nothing here is ever given server
// markup, so there is no innerHTML anywhere in this file either.
var CSS =
  ".b{position:fixed;right:12px;bottom:12px;z-index:2147483647;" +
  "font:12px/1.4 ui-monospace,SFMono-Regular,Menlo,monospace;" +
  "background:#1c1917;color:#fafaf9;border:1px solid #44403c;border-radius:6px;" +
  "padding:6px 10px;box-shadow:0 2px 8px rgba(0,0,0,.35);opacity:.92}" +
  ".d{display:inline-block;width:7px;height:7px;border-radius:50%;" +
  "background:#f59e0b;margin-right:7px;vertical-align:middle}";

var host = null;
var label = null;

function badge() {
  if (host) return;
  host = document.createElement("div");
  host.id = "gotth-live-dev-reload";
  var shadow = host.attachShadow ? host.attachShadow({ mode: "open" }) : host;
  try {
    var sheet = new CSSStyleSheet();
    sheet.replaceSync(CSS);
    shadow.adoptedStyleSheets = [sheet];
  } catch (x) {
    // An unstyled marker is still a marker. A dev tool that needs the page's
    // CSP relaxed to show one line of text is worse than a plain one.
  }
  var box = document.createElement("div");
  box.className = "b";
  var dot = document.createElement("span");
  dot.className = "d";
  label = document.createElement("span");
  box.appendChild(dot);
  box.appendChild(label);
  shadow.appendChild(box);
  document.body.appendChild(host);
}

// paint shows, hides or relabels the badge for one action. It is separate from
// observe() so that the decision has no DOM in it at all, which is what lets
// the specs run under node.
function paint(action) {
  if (action === "linked") {
    if (host && host.parentNode) host.parentNode.removeChild(host);
    host = null;
    label = null;
    return;
  }
  if (!document.body) return;
  badge();
  label.textContent =
    action === "reload" ? "gotth-live: new build — reloading" : "gotth-live: waiting for the server";
}

//#endregion

//#region bootstrap

// The guard is what lets the specs import this module under node, where there
// is no document to read a script tag off and no timer worth starting — the
// same arrangement runtime.js and inspector.js use for the same reason.
if (typeof document !== "undefined") {
  var script = document.currentScript;
  var mount = script && script.getAttribute("data-gotth-dev-url");
  var stamped = script && script.getAttribute("data-gotth-dev-build");

  if (mount) {
    configure({ build: stamped, endpoint: buildURL(mount) });

    // One timer and one request in flight, ever. Both guards are here because
    // the visibilitychange listener below deliberately polls out of turn, and
    // a listener that starts a second loop every time the developer alt-tabs
    // is a dev tool that gets slower the longer it is open.
    var timer = 0;
    var busy = false;

    var schedule = function (ms) {
      if (timer) clearTimeout(timer);
      timer = setTimeout(tick, ms);
    };

    var tick = function () {
      timer = 0;
      // Polling stops while the tab is hidden. A background tab is not
      // watching for a reload, and browsers throttle its timers anyway; the
      // visibilitychange listener is what makes the reload land the moment
      // the developer comes back to it.
      if (document.hidden || busy) return void schedule(STEADY);
      busy = true;
      check(globalThis.fetch).then(function (action) {
        busy = false;
        paint(action);
        if (action === "reload") return void globalThis.location.reload();
        schedule(nextDelay());
      });
    };

    document.addEventListener("visibilitychange", function () {
      if (!document.hidden && !state().reloading) schedule(0);
    });

    schedule(STEADY);
  }
}

//#endregion
